### PR TITLE
Update allowedNamespaces type in AzureClusterIdentity

### LINKED ISSUE #1288: Change multi-tenancy allowed namespaces to be more restrictive
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- Currently in the multi-tenancy implemenations, allowed namespaces is by default open to any namespaces
- Other CAP* provider have a little bit more restrictive approach like the following:
```
empty allowedNamespaces --> allows all namespaces
nil allowedNamespaces --> blocks all namespaces

nil NamespaceList --> blocks all
nil Selector --> blocks all
```
- Consistency between providers is highly desirable, so the request here is to make capz consistent with other providers by  adding LabelSelector and following the behaviour mentioned above 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
discussion here: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1149#discussion_r597226974 

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by nader-ziada**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- Change rules for allowedNamespaces in AzureClusterIdentity to be consistent with other providers
```
empty allowedNamespaces --> allows all namespaces
nil allowedNamespaces --> blocks all namespaces

nil NamespaceList --> blocks all
nil Selector --> blocks all
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1288 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Update AzureClusterIdentity to be more restrictive with the following rules
  - empty allowedNamespaces --> allows all namespaces
  - nil allowedNamespaces --> blocks all namespaces
  - nil NamespaceList --> blocks all
  - nil Selector --> blocks all
```

### CODE REVIEW

File: api/v1alpha4/azureclusteridentity_types.go [hunk 1]
```diff
@@ -36,14 +54,15 @@ type AzureClusterIdentitySpec struct {
 	ClientSecret corev1.SecretReference `json:"clientSecret,omitempty"`
 	// Service principal primary tenant id.
 	TenantID string `json:"tenantID"`
-	// AllowedNamespaces is an array of namespaces that AzureClusters can
-	// use this Identity from.
+	// AllowedNamespaces is used to identify the namespaces the clusters are allowed to use the identity from.
+	// Namespaces can be selected either using an array of namespaces or with label selector.
+	// An empty allowedNamespaces object indicates that AzureClusters can use this identity from any namespace.
+	// If this object is nil, no namespaces will be allowed (default behaviour, if this field is not provided)
+	// A namespace should be either in the NamespaceList or match with Selector to use the identity.
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply api/v1alpha4/azureclusteridentity_types.go:61]: good point, will add in desc here and docs

File: api/v1alpha3/azureclusteridentity_conversion.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		return err
+	}
+
+	// Manually restore data.
+	restored := &infrav1alpha4.AzureClusterIdentity{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	if len(src.Spec.AllowedNamespaces) > 0 {
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply api/v1alpha3/azureclusteridentity_conversion.go:39]: restored is the v1alpha4 type

File: api/v1alpha3/azureclusteridentity_conversion.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return err
+	}
+
+	// Preserve Hub data on down-conversion.
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return err
+	}
+	
+	if len(src.Spec.AllowedNamespaces.NamespaceList) > 0 {
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply api/v1alpha3/azureclusteridentity_conversion.go:66]: it wasn't a pointer at first then forget to change this after I made it a pointer

**Comment by nader-ziada** [comment]
@shysank please take another look

File: api/v1alpha3/azureclusteridentity_conversion.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shysank) [reply api/v1alpha3/azureclusteridentity_conversion.go:39]: IIUC, we must restore the extra fields (selector in this case) that will get lost in down conversion here, right?

**Comment by shysank** [comment]
lgtm with one clarification on the conversion logic.

File: api/v1alpha3/azureclusteridentity_conversion.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nader-ziada) [reply api/v1alpha3/azureclusteridentity_conversion.go:39]: you mean in an annotation? doesn't it get saved automatically?

[the remark this answers is not in the document]
Comment (shysank) [reply api/v1alpha3/azureclusteridentity_conversion.go:39]: yeah from the data annotation, something like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/f8b2304ce244e60c8b8c4c6eb78f211e91a0e031/api/v1alpha3/azurecluster_conversion.go#L52 ?

[the remark this answers is not in the document]
Comment (nader-ziada) [reply api/v1alpha3/azureclusteridentity_conversion.go:39]: oh i get it now, my bad, just fixed it

File: azure/scope/identity.go [hunk 4]
```diff
@@ -161,3 +163,45 @@ func getAzureIdentityType(identity *infrav1.AzureClusterIdentity) (aadpodv1.Iden
 	return 0, errors.New("AzureIdentity does not have a vaild type")
 
 }
+
+// IsClusterNamespaceAllowed indicates if the cluster namespace is allowed
+func IsClusterNamespaceAllowed(ctx context.Context, k8sClient client.Client, allowedNamespaces *infrav1.AllowedNamespaces, namespace string) bool {
```
Comment (nader-ziada) [inline azure/scope/identity.go:168]: @devigned can you take a look at this and see if it makes sense based on all the previous conversations?

**Review by devigned — APPROVED** [review]
/lgtm

@CecileRobertMichon or @shysank any other feedback?

**Comment by nader-ziada** [comment]
did a rebase because the check jobs were stuck, will need a new lgtm :)

**Pull request merged**